### ISSUE TITLE
[HOLD for payment 2024-10-11][$500] DeleteComment not working as expected

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#39432]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue reported by:** @shubham1206agra @rayane-djouah 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1711678317008389?thread_ts=1711571474.268739&cid=C01GTK53T8Q

## Action Performed:
1. Add a comment
2. Create a thread in the same comment
3. Add another comment inside the thread
4. Try to delete the comment in the thread
5. Observe that the `DeleteComment` API never runs inside the network tab
6. Try to go to the same thread after navigating away

## Expected Result:
The comment deleted in step 4 should be gone.

## Actual Result:
The comment deleted in step 4 remains.

## Workaround:
n/a

## Platforms:
Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://expensify.slack.com/archives/C01GTK53T8Q/p1712034411209499?thread_ts=1711571474.268739&cid=C01GTK53T8Q

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014a6349d3b575268d</li>
        <li>Upwork Job ID: 1775247417597710336</li>
        <li>Last Price Increase: 2024-04-02</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Contributor | 0</li>
<li>bernhardoj | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @slafortune</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Delete comment request is never processed.

### What is the root cause of that problem?
We include delete comment as part of the conflicting request and `shouldIncludeCurrentRequest` is also true (except thread parent)
https://github.com/Expensify/App/blob/879378ff68dc026c1b4ba6af30b70051b99c6d35/src/libs/actions/Report.ts#L1258-L1267

so when we do a delete comment, it conflicts with itself and is removed from the request list.

### What changes do you think we should make in order to solve the problem?

From my understanding, the purpose of `getConflictingRequests` is:
1. Remove duplicate requests, such as reconnect app (which is achieved previously with idem key)
2. Remove both add (optionally edit) and delete request of a same action as they cancel out each other

The 1st point works fine, but not with the 2nd. There are 2 issues with the 2nd point:
1. The delete comment is always removed from the request list, but it should only be removed when there is an add request.
2. If we add a comment while online, then edit and delete it while offline, both edit and delete command is removed from the request list and we optimistically remove the message.

So, here is what I propose:
1. Replace `shouldIncludeCurrentRequest` with `shouldExcludeCurrentRequestWhenConflict`, the value stays the same
2. Remove DELETE from the conflicting request list and conditionally add UPDATE if the `action.isOptimisticAction` is true
https://github.com/Expensify/App/blob/879378ff68dc026c1b4ba6af30b70051b99c6d35/src/libs/actions/Report.ts#L1262
3. In `PersistedRequest.save`, don't include the current request (`requestToPersist`) to the array yet, this means we don't need `potentiallyConflictingRequests` anymore. We will pass `requests` to `getConflictingRequests`
https://github.com/Expensify/App/blob/879378ff68dc026c1b4ba6af30b70051b99c6d35/src/libs/actions/PersistedRequests.ts#L20-L21
4. Create a new variable called `hasConflict` with an initial value of false.
5. Set it to true if `conflictingRequests` is not empty
6. If `shouldExcludeCurrentRequestWhenConflict` is false or there is no conflict, add the current request to the request list

<details>
<summary>diff</summary>

```diff
diff --git a/src/libs/actions/PersistedRequests.ts b/src/libs/actions/PersistedRequests.ts
index 8adb598246..7e671e8cce 100644
--- a/src/libs/actions/PersistedRequests.ts
+++ b/src/libs/actions/PersistedRequests.ts
@@ -18,16 +18,15 @@ function clear() {
 }
 
 function save(requestToPersist: Request) {
-    const requests = [...persistedRequests, requestToPersist];
+    const requests = [...persistedRequests];
+    let hasConflict = false;
 
     // identify and handle any existing requests that conflict with the new one
-    const {getConflictingRequests, handleConflictingRequest, shouldIncludeCurrentRequest} = requestToPersist;
+    const {getConflictingRequests, handleConflictingRequest, shouldExcludeCurrentRequestWhenConflict} = requestToPersist;
     if (getConflictingRequests) {
-        // Get all the requests, potentially including the one we're adding, which will always be at the end of the array
-        const potentiallyConflictingRequests = shouldIncludeCurrentRequest ? requests : requests.slice(0, requests.length - 1);
-
         // Identify conflicting requests according to logic bound to the new request
-        const conflictingRequests = getConflictingRequests(potentiallyConflictingRequests);
+        const conflictingRequests = getConflictingRequests(requests);
+        hasConflict = conflictingRequests.length > 0;
 
         conflictingRequests.forEach((conflictingRequest) => {
             // delete the conflicting request
@@ -41,11 +40,15 @@ function save(requestToPersist: Request) {
         });
     }
 
+    if (!shouldExcludeCurrentRequestWhenConflict || !hasConflict) {
+        requests.push(requestToPersist)
+    }
+
     // Don't try to serialize conflict resolution functions
     persistedRequests = requests.map((request) => {
         delete request.getConflictingRequests;
         delete request.handleConflictingRequest;
-        delete request.shouldIncludeCurrentRequest;
+        delete request.shouldExcludeCurrentRequestWhenConflict;
         return request;
     });
 
diff --git a/src/libs/actions/Report.ts b/src/libs/actions/Report.ts
index 775d2459b0..0ab400d523 100644
--- a/src/libs/actions/Report.ts
+++ b/src/libs/actions/Report.ts
@@ -1314,12 +1314,12 @@ function deleteReportComment(reportID: string, reportAction: ReportAction) {
                 const conflictingCommands = (
                     isDeletedParentAction
                         ? [WRITE_COMMANDS.UPDATE_COMMENT]
-                        : [WRITE_COMMANDS.ADD_COMMENT, WRITE_COMMANDS.ADD_ATTACHMENT, WRITE_COMMANDS.UPDATE_COMMENT, WRITE_COMMANDS.DELETE_COMMENT]
+                        : [WRITE_COMMANDS.ADD_COMMENT, WRITE_COMMANDS.ADD_ATTACHMENT, reportAction.isOptimisticAction ? WRITE_COMMANDS.UPDATE_COMMENT : '']
                 ) as string[];
                 return persistedRequests.filter((request) => conflictingCommands.includes(request.command) && request.data?.reportActionID === reportActionID);
             },
             handleConflictingRequest: () => Onyx.update(successData),
-            shouldIncludeCurrentRequest: !isDeletedParentAction,
+            shouldExcludeCurrentRequestWhenConflict: !isDeletedParentAction,
         },
     );
 }

```
</details>

**Comment by roryabraham** [comment]
next step - waiting for an approved proposal

**Comment by bernhardoj** [comment]
I have checked and it's a different issue unrelated to the conflict resolution logic. I have posted a proposal there explaining the issue.

**Comment by shahinyan11** [comment]
@bernhardoj's solution looks good. But I want us to pay attention to one thing. Here we are comparing two objects using the `===` operator. Are we sure that object references will never be changed at this step? @bernhardoj @roryabraham @shubham1206agra  Are you think it is a reliable implementation ?

**Comment by grgia** [comment]
@roryabraham I'm going to unassign as it looks like you've been on top this one, feel free to reassign if you need me as engineer review!

**Comment by roryabraham** [comment]
reviewed the PR

**Comment by mallenexpensify** [comment]
Experiencing this bug, same thing?
- post a comment in a room and a pic in the same room (not thread)
- delete comment and delete pic
- Navigate to another chat
- Return to the room
- Obverse deleted comment and pic weren't deleted. 

I have a vid if needed/wanted

**Comment by roryabraham** [comment]
Taking this internal over [here](https://github.com/Expensify/App/pull/40059) for urgency, but @bernhardoj should still be paid since they provided much of the solution and opened a PR, which might have been merged already if I had reviewed faster.

**Comment by roryabraham** [comment]
No worries, I don't think you or @bernhardoj did anything wrong - you've both been great. ❤️  I just decided to bring this internal because we have a fireroom. I agree you should both be compensated fully, and I'd appreciate your help reviewing the new PR to ensure that it ends up fixing the problem(s) without introducing new ones. :teamwork:

**Comment by slafortune** [comment]
@bernhardoj paid $500 ✅ 
@shubham1206agra can you please complete the checklist and accept this [offer](https://www.upwork.com/nx/wm/offer/101716355)

**Comment by bernhardoj** [comment]
Thanks for the payment, but are we supposed to get paid already? I thought we still wait for https://github.com/Expensify/App/pull/40059?

**Comment by slafortune** [comment]
Ah shoot!  you are correct @bernhardoj  - let's let this sit for the time being rather than undoing and redoing!

**Comment by slafortune** [comment]
Still waiting!

**Comment by roryabraham** [comment]
I actually think we can pay out @bernhardoj, but @shubham1206agra is still on the hook for the review of https://github.com/Expensify/App/pull/40059. Sorry for the delay here, was working on a few more urgent things. Let's pay out @bernhardoj then take this off HOLD. I'll resume https://github.com/Expensify/App/pull/40059 ASAP.

**Comment by slafortune** [comment]
Alrighty - @bernhardoj you have been paid and @shubham1206agra you have NOT been paid.  So things are just as expected now.  I will update the title!

**Comment by slafortune** [comment]
@roryabraham @shubham1206agra - the PR looks real close - do you have an update?

**Comment by roryabraham** [comment]
Sorry, I really _should_ have an update but just got pulled into a number of different issues this week. This one is still on my radar and I'll try to get the PR completed next week

**Comment by roryabraham** [comment]
This issue is actually reasonably high priority because it relates to reducing server traffic by removing redundant requests and improving app performance. However, I'm focusing more on https://github.com/Expensify/App/issues/42214 right now

**Comment by slafortune** [comment]
Are you able to find someone to reassign this to @roryabraham ?  I will be on vacation until June 4th, so adding another BZ in the meantime.  I'll take this back on the 4th.

**Comment by adelekennedy** [comment]
same issue here - it seems like this needs to be an internal issue but @roryabraham is focusing on another high priority issue. @roryabraham can this be moved to `Hot Picks` so someone else can pick this up? Or is there someone you would recommend I reassign to?

**Comment by roryabraham** [comment]
So here's the status here - the original bug reported here is no longer reproducible. We "fixed" it by removing what's ultimately just a performance optimization that de-dupes network requests. For example, if you're offline and you edit a comment then delete it, when you come back online it only makes sense to delete it, rather than editing and deleting it, because the end result is the same

In [this PR](https://github.com/Expensify/App/pull/40059), I started working on adding that performance optimization back, ideally without introducing any new bugs.

So I think the issue is at least medium priority because performance is a critical topic right now. But TBC, this isn't a major bottleneck - it only affects the UX when you go from Offline -> Online.

I could potentially reassign this issue and pass off the [WIP PR](https://github.com/Expensify/App/pull/40059) to someone else, but I'm not sure that this is so critical that that is warranted right now. It's also generally a high-risk change and I've got a lot of context on the code and the bugs that we've experienced changing it in the past. So my recommendation is that I hold onto this for now, but we can make it a weekly.

**Comment by adelekennedy** [comment]
thank you! That makes sense to me

**Comment by slafortune** [comment]
Thanks @adelekennedy @roryabraham - I'll take this back over now.

**Comment by slafortune** [comment]
@roryabraham should this still be left to weekly?

**Comment by roryabraham** [comment]
No update, this one isn't really urgent. Going to leave it as a weekly for now though

**Comment by roryabraham** [comment]
Finally resuming work on [the PR](https://github.com/Expensify/App/pull/40059), just resolved conflicts and am going to continue with more automated tests

**Comment by slafortune** [comment]
This is waiting on Rory, so I don't feel the need to add another BZ person to this.
I'll be out until 8/21 and will check in on this then.

**Comment by slafortune** [comment]
@roryabraham I am not able to recreate this any longer - following the steps above I was able to successfully delete the comment on the thread - still gone after navigating away and back.
<img width="597" alt="image" src="https://github.com/user-attachments/assets/fa6174c2-b707-471d-ae2a-ffc60137b638">

**Comment by slafortune** [comment]
Whoops, I didn't mean to close, I meant to clarify that this isn't reproducible and that we are still waiting on the linked PR to finalize payments as noted here - https://github.com/Expensify/App/issues/39432#issuecomment-2092447582

**Comment by roryabraham** [comment]
We've been actively developing/reviewing in https://github.com/Expensify/App/pull/47913, seems we are close

**Comment by roryabraham** [comment]
https://github.com/Expensify/App/pull/47913 was deployed to prod today. Not sure why this didn't get marked as awaiting payment, doing that manually

**Comment by slafortune** [comment]
@shubham1206agra Role C+ paid $500 vai upworks
@bernhardoj Role Contributor had [already been paid](https://github.com/Expensify/App/issues/39432#issuecomment-2078454415) $500
